Add support for generating CS defines for SPI GPIO CS

galak opened the pull request.

galak committed:
scripts/dts/extract: Move extract_{controller,cells} into globals

Move extract_cells and extract_controller functions into globals so they
can be utilized outside of just the main extract_dts_includes script.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts/dts/extract: Add support for generating SPI GPIO CS defines

Add support to generation script to generate defines for GPIO CS
[WIP], need to clarify some issues on how CS's look when we mix GPIO-CS
and HW CS.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

pabigot reviewed (dismissed):
I'm not sure this is going the right way, but it may be because I'm looking at an unusual target.

There isn't in-tree SPI stuff I'm familiar with, so this is using my WIP on spi-nor on Nordic (nrf52840_pca10056 and particle_xenon).  With this in the DTS:

```
&spi2 { /* dedicated MX25L */
	status = "ok";
	sck-pin = <19>;
	mosi-pin = <20>;
	miso-pin = <21>;
	cs-gpios = <&gpio0 17 0>;
	gd25q32c@0 {
		compatible = "jedec,spi-nor";
		label = "GD25Q32C";
		reg = <0>;
		spi-max-frequency = <80000000>;
		wp-gpios = <&gpio0 22 0>;
		hold-gpios = <&gpio0 23 0>;
		jedec-id = <0xc2 0x20 0x16>;
	};
};
```
the added defines with your branch produce:
```
 #define DT_NORDIC_NRF_SPI_40023000_CS_GPIOS_CONTROLLER "GPIO_0"
 #define DT_NORDIC_NRF_SPI_40023000_CS_GPIOS_FLAGS      0
 #define DT_NORDIC_NRF_SPI_40023000_CS_GPIOS_PIN        17
...
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BASE_ADDRESS          0x0
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BUS_NAME              "SPI_2"
+#define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER    "GPIO_0"
+#define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS         0
+#define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN           17
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_CONTROLLER "GPIO_0"
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_FLAGS      0
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_PIN        23
...
 #define DT_JEDEC_SPI_NOR_0_BASE_ADDRESS          DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BASE_ADDRESS
 #define DT_JEDEC_SPI_NOR_0_BUS_NAME              DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BUS_NAME
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER    DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS         DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN           DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN
 #define DT_JEDEC_SPI_NOR_0_HOLD_GPIOS_CONTROLLER DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_CONTROLLER
 #define DT_JEDEC_SPI_NOR_0_HOLD_GPIOS_FLAGS      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_FLAGS
 #define DT_JEDEC_SPI_NOR_0_HOLD_GPIOS_PIN        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_PIN
```
But the whole point of having chip-selects on the SPI bus *should* be that the GPIOs are associated with the bus, and the device just uses an index.  So it should be generating something like:
```
 #define DT_NORDIC_NRF_SPI_40023000_CS_GPIOS_CONTROLLER "GPIO_0"
 #define DT_NORDIC_NRF_SPI_40023000_CS_GPIOS_FLAGS      0
 #define DT_NORDIC_NRF_SPI_40023000_CS_GPIOS_PIN        17
+#define DT_NORDIC_NRF_SPI_40023000_CS_GPIOS { {"GPIO_0", 0, 17} }
+#define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER    "GPIO_0"
+#define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS         0
+#define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN           17
...
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BASE_ADDRESS          0x0
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BUS_NAME              "SPI_2"
+#define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO               0
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_CONTROLLER "GPIO_0"
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_FLAGS      0
 #define DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_PIN        23
...
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO
```

Here I'm proposing that we pack the cs-gpios into something that can be used in the driver initialization, e.g.:
```
static const spi_cs_control_cfg spi0_cs_cfgs[] = DT_NORDIC_NRF_SPI_0_CS_GPIOS;
```
then the SPI devices using something like:
```
static const struct spi_config dev0 = {
   .slave = DT_JEDEC_SPI_NOR_0_CS_GPIO,
};
```
I do understand that this means the Nordic SPI driver has to be updated to follow other devices in actually using the cs index rather than the full cs tuple, but that should happen anyway.  I assume that targets where there's a dedicated CS gpio would also want the index.

Potentially related, if there are multiple entries in the `cs_gpios` property value this is producing:
```
 #define DT_JEDEC_SPI_NOR_0_BASE_ADDRESS          DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BASE_ADDRESS
 #define DT_JEDEC_SPI_NOR_0_BUS_NAME              DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BUS_NAME
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_0  DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_0
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1  DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_0       DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_0
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1       DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN_0         DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_0
+#define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1         DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1
 #define DT_JEDEC_SPI_NOR_0_HOLD_GPIOS_CONTROLLER DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_CONTROLLER
 #define DT_JEDEC_SPI_NOR_0_HOLD_GPIOS_FLAGS      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_FLAGS
 #define DT_JEDEC_SPI_NOR_0_HOLD_GPIOS_PIN        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_HOLD_GPIOS_PIN
```
which is not helpful at all.

*ADD* For the last I expect it's `DT_JEDEC_SPI_NOR_0_BASE_ADDRESS` that's to be used as the "slave" index, but if we can use that we don't need all the detailed ones.  See also https://github.com/zephyrproject-rtos/zephyr/issues/12226#issuecomment-456587539

lgl88911 commented on scripts/dts/extract/reg.py:
```diff
@@ -28,12 +28,36 @@ def extract(self, node_address, names, def_label, div):
 
         (nr_address_cells, nr_size_cells) = get_addr_size_cells(node_address)
 
+        if 'parent' in binding:
```
The flow mean: Son will use parent description and create MACRO? Eg follow code: the enc28j60 own cs-gpios ?
```
&spi1 {
	status = "ok";
	cs-gpios = <&gpio0 7 0>;

	enc28j60@0 {

	};
};
```

galak replied:
@lgl88911 current the CS information should be specified at the controller node level in dts, not at the SPI device level.

lgl88911 replied:
Base the pr , flow DS is right? If it right, the dts python script will handle the change and no need modify the dts binding yaml (like #12641)
&spi1 {
	status = "ok";
	enc28j60@0 {
             cs-gpios = <&gpio0 7 0>;
	};
};

pabigot replied:
@lgl88911 See #12226.  #12641 is the wrong solution to the problem.

galak commented:
> @galak is there a particular in-tree board and sample you want me to look at so my comments are more relevant to the immediate goal?

You can look at the following boards:

boards/arm/disco_l475_iot1/disco_l475_iot1.dts: cs-gpios = <&gpiod 13 0>, <&gpioe 0 0>;
boards/arm/nrf51_ble400/nrf51_ble400.dts:       cs-gpios = <&gpio0 30 0>;
boards/arm/olimexino_stm32/olimexino_stm32.dts: cs-gpios = <&gpiod 2 0>;
boards/arm/reel_board/reel_board.dts:   cs-gpios = <&gpio0 17 0>;
boards/x86/arduino_101/arduino_101.dts: cs-gpios = <&gpio0 24 0>;
boards/x86/quark_se_c1000_devboard/quark_se_c1000_devboard.dts: cs-gpios = <&gpio0 11 0>;

And the following drivers examples:
drivers/wifi/eswifi
drivers/wifi/winc1500
drivers/ieee802154/*cc2520*

galak committed:
wifi: eswifi: Use generated SPI CS define

Move to using the generated DT_INVENTEK_ESWIFI_ESWIFI0_CS_GPIO_pin and
DT_INVENTEK_ESWIFI_ESWIFI0_CS_GPIO_CONTROLLER defines and drop the
DT_ESWIFI0_CS_GPIOS_PIN & DT_INVENTEK_ESWIFI_ESWIFI0_CS_GPIO_PIN from
dts_fixup.hl

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

pabigot commented:
> So I agree w/you that really the APIs around SPI should change and the handling of GPIOs used for CS should be done in the bus controller driver and not specified in each "slave"/client driver. However, I didn't want to take on changing the API at this time. 

I've gotta say, I'm getting a bit of whiplash trying to figure out how Zephyr architectural decisions are made.

We now agree in #12226 that chip-selects should be associated the bus rather than device because that's what Linux does---even though Zephyr went the other way when the "new" SPI API was added two years ago.  That same SPI API which is one of the exactly **two** APIs Zephyr [claims is "stable"](https://github.com/zephyrproject-rtos/zephyr/projects/15) (the other being ADC, which is [not functionally complete](https://github.com/zephyrproject-rtos/zephyr/issues/11922)).

But instead of either reconsidering the decision to ignore Linux and put `cs-gpios` in the device where Zephyr expects it, or biting the bullet and changing the API, we're going to have the API and DT architectures remain inconsistent.  In an LTS release.

Yes, I am getting a bit frustrated with Zephyr. 

> So generating something like:
> 
> ```
> #define DT_ATMEL_WINC1500_0_CS_GPIO_CONTROLLER						DT_SNPS_DESIGNWARE_SPI_B0001400_ATMEL_WINC1500_0_CS_GPIO_CONTROLLER
> #define DT_ATMEL_WINC1500_0_CS_GPIO_FLAGS						DT_SNPS_DESIGNWARE_SPI_B0001400_ATMEL_WINC1500_0_CS_GPIO_FLAGS
> #define DT_ATMEL_WINC1500_0_CS_GPIO_PIN							DT_SNPS_DESIGNWARE_SPI_B0001400_ATMEL_WINC1500_0_CS_GPIO_PIN
> ```
> means we can replace the following dts_fixup.h:
> 
> ```
> #define DT_WIFI_WINC1500_GPIO_SPI_CS_DRV_NAME   DT_SNPS_DESIGNWARE_SPI_B0001400_CS_GPIOS_CONTROLLER
> #define DT_WIFI_WINC1500_GPIO_SPI_CS_PIN        DT_SNPS_DESIGNWARE_SPI_B0001400_CS_GPIOS_PIN
> ```

OK, that'd be nice.  But in the case where there are multiple entries in the bus `cs-gpios` properties the binding generation must generate an un-suffixed identifier based on the `reg` property.  We can't assume there's only one element in the `cs-gpios` value array, and we can't expect every SPI device driver to have to deal with managing the indexing itself.

galak commented:
> > So I agree w/you that really the APIs around SPI should change and the handling of GPIOs used for CS should be done in the bus controller driver and not specified in each "slave"/client driver. However, I didn't want to take on changing the API at this time.
> 
> I've gotta say, I'm getting a bit of whiplash trying to figure out how Zephyr architectural decisions are made.
> 
> We now agree in #12226 that chip-selects should be associated the bus rather than device because that's what Linux does---even though Zephyr went the other way when the "new" SPI API was added two years ago. That same SPI API which is one of the exactly **two** APIs Zephyr [claims is "stable"](https://github.com/zephyrproject-rtos/zephyr/projects/15) (the other being ADC, which is [not functionally complete](https://github.com/zephyrproject-rtos/zephyr/issues/11922)).
> 
> But instead of either reconsidering that decision and ignoring Linux to put `cs-gpios` in the device where Zephyr expects it, or biting the bullet and changing the API, we're going to have the API and DT architectures remain inconsistent. In an LTS release.
> 
> Yes, I am getting a bit frustrated with Zephyr.

I understand, there are a lot of moving parts in Zephyr and its hard to get everything aligned at once and all different aspects, driver API, dts, etc sync'd at the same time, let alone having the bandwidth for experts in one area to review and comment on another.

> 
> > So generating something like:
> > ```
> > #define DT_ATMEL_WINC1500_0_CS_GPIO_CONTROLLER						DT_SNPS_DESIGNWARE_SPI_B0001400_ATMEL_WINC1500_0_CS_GPIO_CONTROLLER
> > #define DT_ATMEL_WINC1500_0_CS_GPIO_FLAGS						DT_SNPS_DESIGNWARE_SPI_B0001400_ATMEL_WINC1500_0_CS_GPIO_FLAGS
> > #define DT_ATMEL_WINC1500_0_CS_GPIO_PIN							DT_SNPS_DESIGNWARE_SPI_B0001400_ATMEL_WINC1500_0_CS_GPIO_PIN
> > ```
> > means we can replace the following dts_fixup.h:
> > ```
> > #define DT_WIFI_WINC1500_GPIO_SPI_CS_DRV_NAME   DT_SNPS_DESIGNWARE_SPI_B0001400_CS_GPIOS_CONTROLLER
> > #define DT_WIFI_WINC1500_GPIO_SPI_CS_PIN        DT_SNPS_DESIGNWARE_SPI_B0001400_CS_GPIOS_PIN
> > ```
> 
> OK, that'd be nice. But in the case where there are multiple entries in the bus `cs-gpios` properties the binding generation must generate an un-suffixed identifier based on the `reg` property. We can't assume there's only one element in the `cs-gpios` value array, and we can't expect every SPI device driver to have to deal with managing the indexing itself.

So the defines I generated are from the SPI device perspective, I'm extracting the specific `cs-gpios` entry that they device needs.

I agree once all this logic moves into the SPI controller we need to generate something different for the controller to see the total view of `cs-gpios`.

pabigot commented:
> > OK, that'd be nice. But in the case where there are multiple entries in the bus `cs-gpios` properties the binding generation must generate an un-suffixed identifier based on the `reg` property. We can't assume there's only one element in the `cs-gpios` value array, and we can't expect every SPI device driver to have to deal with managing the indexing itself.
> 
> So the defines I generated are from the SPI device perspective, I'm extracting the specific `cs-gpios` entry that they device needs.

Not in the case of multiple cs-gpios at the bus level:
```
#define DT_JEDEC_SPI_NOR_0_BASE_ADDRESS         DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BASE_ADDRESS
#define DT_JEDEC_SPI_NOR_0_BUS_NAME             DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BUS_NAME
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_0 DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_0
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1 DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_0      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_0
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN_0        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_0
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1
```
This really has to be generated as:
```
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1
#define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1
```
when the device is `reg = 1`.

galak commented:
> Not in the case of multiple cs-gpios at the bus level:
> 
> ```
> #define DT_JEDEC_SPI_NOR_0_BASE_ADDRESS         DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BASE_ADDRESS
> #define DT_JEDEC_SPI_NOR_0_BUS_NAME             DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_BUS_NAME
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_0 DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_0
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1 DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_0      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_0
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN_0        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_0
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1
> ```
> This really has to be generated as:
> 
> ```
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_CONTROLLER_1
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS      DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_FLAGS_1
> #define DT_JEDEC_SPI_NOR_0_CS_GPIO_PIN        DT_NORDIC_NRF_SPI_40023000_JEDEC_SPI_NOR_0_CS_GPIO_PIN_1
> ```
> when the device is `reg = 1`.

Weird, I'm not seeing this with the disco_l475_iot1 which has:

```
&spi3 {
        status = "ok";

        cs-gpios = <&gpiod 13 0>, <&gpioe 0 0>;
```

galak commented:
@pabigot do you have a tree that I can see your jedec example and test on?

pabigot commented:
I updated my comment to include the elided material from cmake output, so you can look for clues there.  Possibly also relevant is my environment setup:
```
export ZEPHYR_TOOLCHAIN_VARIANT=gnuarmemb
export GNUARMEMB_TOOLCHAIN_PATH=/usr/local/gcc-arm-none-eabi
export PATH=${HOME}/.local/bin:${PATH}
cd /mnt/devel/external/zephyr
. zephyr-env.sh
```
Possibly also relevant is that anything in `${HOME}/.local/bin` would have been whatever `pip` magic I had to run back in late November to get started with Zephyr, and that I don't have any special west versions anywhere. But I do have:
```
lilith[621]$ dtc --version
Version: DTC 1.4.6
```

pabigot commented:
Yep, it's the `dtc` version.  If I update to 1.4.7 it works.

Whatever.

galak commented:
> Yep, it's the `dtc` version. If I update to 1.4.7 it works.
> 
> Whatever.

I was guessing it was dtc version.  Can you send me the result of:
`grep cs-gpios zephyr/nrf52840_pca10056.dts_compiled `

pabigot commented:
That's a clue.  With 1.4.6 it's:
```
			cs-gpios = <0x2 0x11 0x0 0x3 0x5 0x0>;
```
With 1.4.7 it's:
```
			cs-gpios = < 0x02 0x11 0x00 >, < 0x03 0x05 0x00 >;
```

galak committed:
scripts/dts/extract: Add single extract to extract_{controller,cells}

Add the ability to extract just a single instance from extract_cells and
extract_controller.  This lets us pass in a prop_values like:

cs-gpios =  < 0x02 0x11 0x00 >, < 0x03 0x05 0x00 >;

and extract just index = 1 if we want < 0x03 0x05 0x00 >.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

pabigot reviewed (approved):
The generated material appears to be sufficient to handle both the current situation (device holds GPIO data extracted from bus) and the currently envisioned future rework to remove that responsibility from the device implementations.
